miden-lib/asm/miden/kernels/tx/tx.masm

```diff
@@ -107,10 +116,144 @@ proc.increment_num_created_notes
     # => [note_idx]
 end
 
+#! Adds a non-fungible asset to a note at the next available position. 
+#! Returns the pointer to the note the asset was stored at.
+#! Panics if the non-fungible asset already exists in the note.
+#!
+#! Inputs: [ASSET, note_ptr, num_of_assets]
+#! Outputs: [note_ptr]
+proc.add_non_fungible_asset_to_note
+    dup.4 exec.memory::get_created_note_asset_data_ptr
+    # => [asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the pointer at which we should stop iterating
+    dup dup.7 add
+    # => [end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # pad the stack and move the pointer to the top
+    padw movup.5
+    # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the loop latch
+    dup dup.6 neq
+    # => [latch, asset_ptr, 0, 0, 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    while.true
+        dup movdn.6
+        # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+        # load the asset and compare
+        mem_loadw push.0 movdn.4 push.0 movdn.4 dupw.2
+        # => [ASSET, ASSET', 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # compare the ASSETs
+        eqw assertz.err=ERR_NON_FUNGIBLE_ASSET_ALREADY_EXISTS
+        # => [ASSET, ASSET', 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # drop ASSETs and increment the asset pointer
+        dropw dropw push.0.0 movup.5 add.1
+        # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # check if we reached the end of the loop
+        dup dup.6 neq
+    end
```

Comment:
I think the state of the stack in the comment on line 139 is off: there shouldn't be `asset_ptr` following `end_asset_ptr`.

But also, I think we can make this a bit more efficient if we enter the loop with something like this:
```
[asset_ptr, 0, 0, 0, 0, ASSET, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]
```
Then, we could immediately execute:
```
mem_loadw eqw assertz.err=ERR_NON_FUNGIBLE_ASSET_ALREADY_EXISTS
#=> [STORED_ASSET, ASSET, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]
```

And then to prepare for the next iteration we could do:
```
movup.9 add.1 dup movdn.10
#=> [asset_ptr+1, STORED_ASSET, ASSET, end_asset_ptr, asset_ptr+1, note_ptr, num_of_assets]

dup.9 dup.1 neq
#=> [latch, asset_ptr+1, STORED_ASSET, ASSET, end_asset_ptr, asset_ptr+1, note_ptr, num_of_assets]
```

And we can skip a bunch of `dropw` and `dupw` operations.

Reply:
Nice, I have it almost like this. I think we should get rid of the STORED_ASSET